com.unity.render-pipelines.universal/Runtime/RendererFeatures/ScreenSpaceShadows.cs

```diff
@@ -0,0 +1,206 @@
+                    return;
+                }
+
+                Camera camera = renderingData.cameraData.camera;
+                ShadowData shadowData = renderingData.shadowData;
+
+                int cascadesCount = shadowData.mainLightShadowCascadesCount;
+
+                CommandBuffer cmd = CommandBufferPool.Get();
+                using (new ProfilingScope(cmd, ProfilingSampler.Get(URPProfileId.ResolveShadows)))
+                {
+                    cmd.SetViewProjectionMatrices(Matrix4x4.identity, Matrix4x4.identity);
+                    cmd.DrawMesh(RenderingUtils.fullscreenMesh, Matrix4x4.identity, material);
+                    cmd.SetViewProjectionMatrices(camera.worldToCameraMatrix, camera.projectionMatrix);
+
+                    CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadows, false);
+                    CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadowsCascades, false);
+                    CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadowsScreen, true);
+                }
+
+                context.ExecuteCommandBuffer(cmd);
+                CommandBufferPool.Release(cmd);
+            }
+
+            /// <inheritdoc/>
+            public override void OnCameraCleanup(CommandBuffer cmd)
+            {
+                if (cmd == null)
+                {
+                    throw new ArgumentNullException("cmd");
+                }
+
+                cmd.ReleaseTemporaryRT(m_RenderTarget.id);
+            }
+        }
+
+        private class RestoreShadowKeywordsPass : ScriptableRenderPass
+        {
+            const string m_ProfilerTag = "Restore Shadow Keywords Pass";
+            private ProfilingSampler m_ProfilingSampler = new ProfilingSampler(m_ProfilerTag);
+
+            public override void Execute(ScriptableRenderContext context, ref RenderingData renderingData)
+            {
+                CommandBuffer cmd = CommandBufferPool.Get();
+                using (new ProfilingScope(cmd, m_ProfilingSampler))
+                {
+                    ShadowData shadowData = renderingData.shadowData;
+                    int cascadesCount = shadowData.mainLightShadowCascadesCount;
+
+                    bool mainLightShadows = renderingData.shadowData.supportsMainLightShadows;
+                    bool receiveShadowsNoCascade = mainLightShadows && cascadesCount == 1;
+                    bool receiveShadowsCascades = mainLightShadows && cascadesCount > 1;
+
+                    // Before transparent object pass, force screen space shadow for main light to disable
+                    CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadowsScreen, false);
+
+                    // then enable main light shadows with or without cascades
+                    CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadows, receiveShadowsNoCascade);
+                    CoreUtils.SetKeyword(cmd, ShaderKeywordStrings.MainLightShadowsCascades, receiveShadowsCascades);
```

Comment:
Not an issue but minor suggestion, restoring keywords could also be done in ScriptableRenderer before rendering opaques block. Othewise by default this pass will switch to Camera target and might cause a render target switch for no reason if the previous pass is not rendering to camera target.

Reply:
It restored those shadow keywords after rendering opaque block for preventing transparent objects from screen space shadow. Those processes can be in transparent setting pass, but there was a issue in UniversalRenderingExamples reported by @cinight .

That sample added renderer feature for FPS camera, this feature was injected after rendering opaque objects. but my restoring keywords was set before transparent pass. the screen space shadow would be incorrect for fps gun space. Because the project matrix was not matched between world space and fps gun space. So I can't avoid making another pass for restoring pass and injecting the pass after rendering opaque pass.